### PR TITLE
db/utils: perform signature validation on original message bytes

### LINKED ISSUE #1459: Some valid PGP-signed emails fail to validate: "Bad signature"
**Describe the bug**

Signed MIME emails sent by some other MUAs such as ProtonMail or K9-Mail (Android) do not validate in alot despite being valid in other MUAs like Evolution.

**Software Versions**
- Python version: 3.7
- Notmuch version: [if relevant]
- Alot version: master / f1cecca

**To Reproduce**
Steps to reproduce the behaviour:
1. import the attached .eml in the mailidr and refresh the notmuch index
2. import the corresponding public key in the gpg keyring with the right trust level
3. open the message in alot, notice that the validation fails: `PGP-Signature Invalid: ...: Bad signature for key(s)`

Example of valid .eml which fails to validate, with the public key file:
https://gist.github.com/pacien/422dd56c5c46967e642177bfbe6390af#file-1575332405-m647013610p6859q2rd52e5c50-eml

This .eml is shown as valid in Evolution.

**Error Log**
/

### PR DESCRIPTION
**Pull request by pacien**
Parsing and re-serialising a Message sometimes alters the original content
(by removing extra lines between nested message multiparts for example).
This interfers with the signature validation which requires the byte content
to match perfectly.

To avoid this issue, we perform the signature validation on the original
message bytes directly.

GitHub: resolves #1459

### CODE REVIEW

File: tests/db/test_utils.py [hunk 1]
```diff
@@ -570,8 +585,24 @@ def test_signed_in_multipart_mixed(self):
...
+        """It is valid to encapsulate a multipart/signed payload inside a
+        multipart/mixed payload, verify that works.
+
+        The signature being sensitive to the way the multipart message
+        was assembled (with blank lines between parts, ...),
+        we need to make sure that we're able to validate messages generated
+        by other MUAs as well.
+        """
+        with open('tests/static/mail/protonmail-signed.eml', 'rb') as f:
+            mb = f.read()
```
Comment (roosemberth) [inline tests/db/test_utils.py:603]: Consider using [pathlib](https://docs.python.org/3/library/pathlib.html).
```suggestion
        mb = Path('tests/static/mail/protonmail-signed.eml', 'rb').read_bytes()
```

File: alot/db/utils.py [hunk 2]
```diff
@@ -110,33 +116,47 @@ def _handle_signatures(original, message, params):
...
-                message.get_payload(0).as_bytes(policy=email.policy.SMTP),
-                message.get_payload(1).get_payload(decode=True))
-        except GPGProblem as e:
-            malformed = str(e)
-
-    add_signature_headers(original, sigs, malformed)
+    try:
+        nb_parts = len(message.get_payload()) if message.is_multipart() else 1
+        if nb_parts != 2:
+            raise MessageError('expected exactly two messages, got {0}'.format(
```
[the remark this answers is not in the document]
Comment (pacien) [reply alot/db/utils.py]: Amended.

File: alot/db/utils.py [hunk 3]
```diff
@@ -110,33 +116,47 @@ def _handle_signatures(original, message, params):
...
+        #  seen a message with 'PGP-'. maybe we should be more permissive here,
+        #  or maybe not, this is crypto stuff...
+        mic_alg = params.get('micalg', 'nothing')
+        if not mic_alg.startswith('pgp-'):
+            raise MessageError('expected micalg=pgp-..., got: {0}'.format(
+                mic_alg))
+
+        # manual part extraction from original message necessary because
+        # parsing and re-serialising a Message isn't byte-perfect,
+        # which interferes with signature validation.
```
[the remark this answers is not in the document]
Comment (pacien) [reply alot/db/utils.py:140]: Amended.

File: tests/db/test_utils.py (hunk 1 above)
Comment (pacien) [reply tests/db/test_utils.py]: Amended.

**Comment by pazz** [comment]
It seems the build only fails for the docs..

**Comment by pacien** [comment]
> It seems the build only fails for the docs..

Not sure of why. It's failing on a file that I didn't touched ([signals.rst](https://travis-ci.com/github/pazz/alot/jobs/316602729#L909)).
The error is the same on [another recent pull request](https://travis-ci.com/github/pazz/alot/jobs/315934756#L908).

**Comment by pazz** [comment]
aha, it seems that travis has updated the sphinx version they use for building docs and that one is less forgiving..

**Comment by pacien** [comment]
It seems that there's another issue with line ending that I forgot to take into account.
This is not ready for merging.

**Comment by pacien** [comment]
The issue with the line ending is solved. This PR is now ready to be reviewed again.

I tested it with some emails sent from ProtonMail (in the unit test) and Evolution and they now validate correctly.

File: alot/db/utils.py [hunk 4]
```diff
@@ -135,10 +135,20 @@ def _handle_signatures(original_bytes, original, message, params):
...
-        # which interferes with signature validation.
-        # See RFC 1847 section 2.1.
+        # RFC 3156 section 5 says that "signed message and transmitted message
+        # MUST be identical". We therefore need to validate against the message
+        # as it was originally sent.
+
+        # The transmitted content and therefore the signed content are using
+        # CRLF as line delimiter, but our eml file has most likely been
+        # converted to UNIX LF line ending in the local storage.
+        if b'\r\n' not in original_bytes:
```
Comment (roosemberth) [inline alot/db/utils.py:145]: Offline discussion: This is horrible, but the RFC3156 does specify the data and headers (including MIME content headers) must be converted [to the canonical <CR><LF> sequence](https://tools.ietf.org/html/rfc3156#section-5). This is unfortunately correct.

**Comment by pacien** [comment]
@pazz up!

**Comment by pazz** [comment]
@pacien please rebase to master. The travis docs issue should be fixed and I want to see the tests passing for all python 3 versions as I'm somewhat worried by the use of email policies (which have been changed around in recent versions).. Cheers,

**Comment by pacien** [comment]
Rebased!

**Pull request merged**